net/nat/ipv6_nat.c

```diff
@@ -0,0 +1,699 @@
+
+#ifdef CONFIG_NET_ICMPv6
+      case IP_PROTO_ICMP6:
+        return ipv6_nat_inbound_icmpv6(ipv6, l4hdr, manip_type);
+#endif
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Name: ipv6_nat_outbound_internal
+ *
+ * Description:
+ *   Check if we want to perform NAT with this outbound packet before
+ *   sending it. If so, translate the local IP/Port to external IP/Port.
+ *
+ * Input Parameters:
+ *   dev        - The device to sent the packet (to get external IP).
+ *   ipv6       - Points to the IPv6 header to translate.
+ *   manip_type - Whether local IP/Port is in source or destination.
+ *
+ * Returned Value:
+ *   The corresponding NAT entry of the packet.
+ *
+ * Assumptions:
+ *   Packet will be sent on NAT device.
+ *
+ ****************************************************************************/
+
+static FAR struct ipv6_nat_entry *
+ipv6_nat_outbound_internal(FAR struct net_driver_s *dev,
+                           FAR struct ipv6_hdr_s *ipv6,
+                           enum nat_manip_type_e manip_type)
+{
+  uint8_t proto;
+  FAR void *l4hdr = net_ipv6_payload(ipv6, &proto);
+
+  switch (proto)
+    {
+#ifdef CONFIG_NET_TCP
+      case IP_PROTO_TCP:
+        return ipv6_nat_outbound_tcp(dev, ipv6, l4hdr, manip_type);
+#endif
+
+#ifdef CONFIG_NET_UDP
+      case IP_PROTO_UDP:
+        return ipv6_nat_outbound_udp(dev, ipv6, l4hdr, manip_type);
+#endif
+
+#ifdef CONFIG_NET_ICMPv6
+      case IP_PROTO_ICMP6:
+        return ipv6_nat_outbound_icmpv6(dev, ipv6, l4hdr, manip_type);
+#endif
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ipv6_nat_inbound
+ *
+ * Description:
+ *   Check if a received packet belongs to a NAT entry. If so, translate it.
+ *
+ * Input Parameters:
+ *   dev   - The device on which the packet is received.
+ *   ipv6  - Points to the IPv6 header with dev->d_buf.
+ *
+ * Returned Value:
+ *   Zero is returned if NAT is successfully applied, or is not enabled for
+ *   this packet;
+ *   A negated errno value is returned if error occured.
+ *
+ ****************************************************************************/
+
+int ipv6_nat_inbound(FAR struct net_driver_s *dev,
+                     FAR struct ipv6_hdr_s *ipv6)
+{
+  /* We only process packets from NAT device and targeting at the address
+   * assigned to the device.
+   */
+
+  if (IFF_IS_NAT(dev->d_flags) &&
+      NETDEV_IS_MY_V6ADDR(dev, ipv6->destipaddr))
+    {
+      FAR struct ipv6_nat_entry *entry =
+          ipv6_nat_inbound_internal(ipv6, NAT_MANIP_DST);
+      if (!entry)
+        {
+          /* Inbound without entry is OK (e.g. towards NuttX itself), skip. */
+
+          return OK;
+        }
+    }
+
+  return OK;
```

Comment:
so we do not have failure case?

Reply:
Yes, copied from IPv4, but seems no situation for failure, removed now.